database/pdb.ksy

```diff
@@ -0,0 +1,961 @@
+  grateful. @GreyCat helped me learn the intricacies (and best
+  practices) of Kaitai far faster than I would have managed on my own.
+
+doc-ref: https://github.com/Deep-Symmetry/crate-digger/blob/master/doc/Analysis.pdf
+
+seq:
+  - contents: [0, 0, 0, 0]
+  - id: len_page
+    type: u4
+    doc: |
+      The database page size, in bytes. Pages are referred to by
+      index, so this size is needed to calculate their offset, and
+      table pages have a row index structure which is built from the
+      end of the page backwards, so finding that also requires this
+      value.
+  - id: num_tables
+    type: u4
+    doc: |
+      Determines the number of table entries that are present. Each
+      table is a linked list of pages containing rows of a particular
+      type.
+  - id: next_unused_page
+    type: u4
+    doc: |
+      @flesinak said: "Not used as any `empty_candidate`, points
+      past the end of the file."
+  - type: u4
+  - id: sequence
+    type: u4
+    doc: |
+      @flesniak said: "Always incremented by at least one,
+      sometimes by two or three."
+  - contents: [0, 0, 0, 0]
+  - id: tables
+    type: table
+    repeat: expr
+    repeat-expr: num_tables
+    doc: |
+      Describes and links to the tables present in the database.
+
+types:
+  table:
+    doc: |
+      Each table is a linked list of pages containing rows of a single
+      type. This header describes the nature of the table and links to
+      its pages by index.
+    seq:
+      - id: type
+        type: u4
+        enum: page_type
+        doc: |
+          Identifies the kind of rows that are found in this table.
+      - id: empty_candidate
+        type: u4
+      - id: first_page
+        type: page_ref
+        doc: |
+          Links to the chain of pages making up that table. The first
+          page seems to always contain similar garbage patterns and
+          zero rows, but the next page it links to contains the start
+          of the meaningful data rows.
+      - id: last_page
+        type: page_ref
+        doc: |
+          Holds the index of the last page that makes up this table.
+          When following the linked list of pages of the table, you
+          either need to stop when you reach this page, or when you
+          notice that the `next_page` link you followed took you to a
+          page of a different `type`.
+    -webide-representation: '{type}'
+
+  page_ref:
+    doc: |
+      An index which points to a table page (its offset can be found
+      by multiplying the index by the `page_len` value in the file
+      header). This type allows the linked page to be lazy loaded.
+    seq:
+      - id: index
+        type: u4
+        doc: |
+          Identifies the desired page number.
+    instances:
+      body:
+        doc: |
+          When referenced, loads the specified page and parses its
+          contents appropriately for the type of data it contains.
```

Comment:
Nice approach in short run (`instance`s are lazy, `seq`s are not), but keep in mind that in long run (when we gonna have everything lazy, https://github.com/kaitai-io/kaitai_struct/issues/65 ) we may want to replace it with indexing of array of pages for better clarity.

Reply:
That is an interesting discussion, when might such a change happen? I may make some changes to the mappings as more things are discovered, for example if we can figure out the index tables, which would make laziness even nicer.

Reply:
Ok, thanks for letting me know about the future potential, but I think we should leave this as it is for now.